### PR TITLE
Fix API server address in cloud-config-download for shoots with disabled DNS

### LINKED ISSUE #4968: Cloud config downloader is missing the API server address with DNS-less seed clusters
**How to categorize this issue?**

/area usability robustness
/kind bug

**What happened**:
I have created a shoot cluster in a landscape where
- There is no internal domain secrets
- The DNS is disabled in the seed settings, `seed.spec.settings.shootDNS.enabled=false`
- It is irrelevant, but I also made sure the gardenlet is started with `APIServerSNI=false`

Effectively, the public IP of the `kube-apiserver` service now has to be used by all clients running outside the seed, e.g. the kubelet. Looking for the value of the server in the OSC resource reveal it is not set there for any of the three OSC downloaders I have:
```bash
$ k get osc -o json | jq  '.items[] | select(.spec.purpose=="provision") | .spec.files[] | select(.path=="/var/lib/cloud-config-downloader/credentials/server")| .content.inline.data | @base64d' -r
https://
https://
https://
```

**What you expected to happen**:
The OSC resource to have the public address of the API server properly set in DNS-less environments.

**How to reproduce it (as minimally and precisely as possible)**:
See above.

**Anything else we need to know?**:

**Environment**:

- Gardener version: 29ffac5ed724458803b7748800e9de6f07c84f0a
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area usability robustness
/kind bug

**What this PR does / why we need it**:

If shoot DNS is disabled, the API server address is not known when the OSC component is instantiated.
With this PR, `ComputeOutOfClusterAPIServerAddress` is called during `DeployOperatingSystemConfig`, when the API server address is known.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/4968

**Special notes for your reviewer**:

Probably broken since https://github.com/gardener/gardener/pull/3454, see https://github.com/gardener/gardener/pull/3454/files#diff-78a67ae80f891c299fbd162c192ef1fe19663c5db0bef7e1a42bb04eb70a6335R68
/invite @rfranzke 

**Release note**:

```other operator
Fixed a bug in the node bootstrapping for shoots without DNS records.
```

### CODE REVIEW

File: pkg/operation/botanist/component/extensions/operatingsystemconfig/operatingsystemconfig.go [hunk 1]
```diff
@@ -66,6 +66,8 @@ type Interface interface {
 	DeleteStaleResources(context.Context) error
 	// WaitCleanupStaleResources waits until all unused OperatingSystemConfig resources are cleaned up.
 	WaitCleanupStaleResources(context.Context) error
+	// SetAPIServerAdress sets the APIServerURL value.
+	SetAPIServerAdress(string)
```
Comment (stoyanr) [inline pkg/operation/botanist/component/extensions/operatingsystemconfig/operatingsystemconfig.go]: There is a typo in the name (`Adress` with single d), but perhaps it's better to use `URL` here to be consistent with the other names?
```suggestion
	SetAPIServerURL(string)
```

**Review by stoyanr — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/component/extensions/operatingsystemconfig/operatingsystemconfig.go (hunk 1 above)
Comment (timebertt) [reply pkg/operation/botanist/component/extensions/operatingsystemconfig/operatingsystemconfig.go]: Good catch, thank you. Fixed in a new commit.
/squash

**Review by stoyanr — CHANGES_REQUESTED** [review]
I tested this a bit with a shoot with a disabled DNS, was able to reproduce the original issue (empty server address in the OSC resource), and fix it by reconciling again with the version from the PR, it all worked well. I only have one rather naming point regarding naming.

**Review by stoyanr — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]
/lgtm

Tested it out locally, thanks!

**Pull request merged**